core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,134 @@
+use std::collections::VecDeque;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_schema_checker_lib::ProtocolSchema;
+
+/// Keeps metadata about receipts stored in outgoing buffer to some shard.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OutgoingBufferMetadata {
+    pub receipt_group_sizes: ReceiptGroupSizes,
+}
+
+impl OutgoingBufferMetadata {
+    pub fn new(group_size_threshold: u32) -> Self {
+        OutgoingBufferMetadata {
+            receipt_group_sizes: ReceiptGroupSizes::V0(ReceiptGroupSizesV0::new(
+                group_size_threshold,
+            )),
+        }
+    }
+
+    /// Update the metadata when a new receipt is added at the end of the buffer.
+    pub fn on_receipt_buffered(&mut self, receipt_size: u64) {
+        match &mut self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.on_receipt_pushed(receipt_size),
+        }
+    }
+
+    /// Update the metadata when a receipt is removed from the front of the buffer.
+    pub fn on_receipt_removed(&mut self, receipt_size: u64) {
+        match &mut self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.on_receipt_popped(receipt_size),
+        }
+    }
+
+    /// Iterate over the sizes of receipts stored in the outgoing buffer.
+    /// Multiple consecutive receipts are grouped into a single group.
+    /// The iterator returns the size of each group, not individual receipts.
+    pub fn iter_receipt_group_sizes(&self) -> impl Iterator<Item = u64> + '_ {
+        match &self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.iter_receipt_group_sizes(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ReceiptGroupSizes {
+    V0(ReceiptGroupSizesV0),
+}
+
+/// Keeps information about the size of receipts stored in the outgoing buffer.
+/// Consecutive receipts are grouped into groups and it keeps aggregate information about the groups,
+/// not individual receipts. Receipts are grouped to keep the size of the struct small.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupSizesV0 {
+    /// Receipts stored in the outgoing buffer, grouped into groups of consecutive receipts.
+    /// Each group stores aggregate information about the receipts in the group.
+    pub groups: VecDeque<ReceiptGroup>,
+    /// A new receipt group is started when the size of the last group exceeds this threshold.
+    /// All groups (except for the first and last one) have at least this size.
+    pub group_size_threshold: u32,
```

Comment:
Does this need to be part of this structure? I suppose you keep it here to support future migrations. Can you explain what are the alternatives? In particular I'm curious about the pros and cons of storing all groups in a single trie node vs storing each in a dedicated trie node.

Reply:
It doesn't, but it was convenient to have it there for the methods that add and remove receipts. The migrations aren't really a concern, we can change the threshold any time we want, it is only used for the last group. The first group is removed when its size reaches zero, the threshold does't matter.

Reply:
This is actually held in two places, here and in the `OutgoingMetadatas`, it seems like a stretch.

Reply:
It needs to be in `OutgoingMetadatas` because I have to pass it to the constructor of some shard's metadata when a new metadata is created. So there are at least two places - `OutgoingMetadatas` and `OutgoingBufferMetadata` where it needs to be.